Source/WebKit/NetworkProcess/EarlyHintsResourceLoader.cpp

```diff
@@ -0,0 +1,141 @@
+    m_hasReceivedEarlyHints = true;
+
+    auto headerValue = response.httpHeaderField(HTTPHeaderName::Link);
+    if (headerValue.isEmpty())
+        return;
+
+    auto url = response.url();
+    ContentSecurityPolicy contentSecurityPolicy { URL { url }, &m_loader, nullptr };
+    contentSecurityPolicy.didReceiveHeaders(ContentSecurityPolicyResponseHeaders { response }, originalRequest().httpReferrer());
+
+    LinkHeaderSet headerSet(headerValue);
+    for (const auto& header : headerSet) {
+        if (!header.valid() || header.url().isEmpty() || header.rel().isEmpty() || header.isViewportDependent())
+            continue;
+
+        if (equalLettersIgnoringASCIICase(header.rel(), "preconnect"_s))
+            startPreconnectTask(response.url(), header, contentSecurityPolicy);
+    }
+}
+
+ResourceRequest EarlyHintsResourceLoader::constructPreconnectRequest(const URL& url)
+{
+    ResourceRequest request { url };
+
+    // firstPartyForCookies and user agent are part of the HTTP socket pool keys in CFNetwork: rdar://59434166
+    auto firstPartyForCookies = originalRequest().firstPartyForCookies();
+    if (firstPartyForCookies.isValid())
+        request.setFirstPartyForCookies(firstPartyForCookies);
+
+    auto userAgent = originalRequest().httpUserAgent();
+    if (!userAgent.isEmpty())
+        request.setHTTPUserAgent(userAgent);
+
+    return request;
+}
+
+void EarlyHintsResourceLoader::startPreconnectTask(const URL& baseURL, const LinkHeader& header, const ContentSecurityPolicy& contentSecurityPolicy)
+{
+    if (!m_loader.parameters().linkPreconnectEarlyHintsEnabled)
+        return;
+
+    URL url(baseURL, header.url());
+    if (!url.isValid() || !url.protocolIsInHTTPFamily())
+        return;
+
+    if (!contentSecurityPolicy.allowConnectToSource(url, ContentSecurityPolicy::RedirectResponseReceived::No, originalRequest().url()))
+        return;
+
+    auto* networkSession = m_loader.connectionToWebProcess().networkSession();
+    if (!networkSession)
+        return;
+
+    NetworkLoadParameters parameters;
+    auto globalFrameID = m_loader.globalFrameID();
+    parameters.webPageProxyID = globalFrameID.webPageProxyID;
+    parameters.webPageID = globalFrameID.webPageID;
+    parameters.webFrameID = globalFrameID.frameID;
+    parameters.storedCredentialsPolicy = equalLettersIgnoringASCIICase(header.crossOrigin(), "anonymous"_s) ? StoredCredentialsPolicy::DoNotUse : StoredCredentialsPolicy::Use;
+    parameters.contentSniffingPolicy = ContentSniffingPolicy::DoNotSniffContent;
+    parameters.contentEncodingSniffingPolicy = ContentEncodingSniffingPolicy::Default;
+    parameters.shouldPreconnectOnly = PreconnectOnly::Yes;
+    parameters.request = constructPreconnectRequest(url);
+    parameters.isNavigatingToAppBoundDomain = m_loader.parameters().isNavigatingToAppBoundDomain;
+    (new PreconnectTask(*networkSession, WTFMove(parameters), [](const WebCore::ResourceError&, const WebCore::NetworkLoadMetrics&) { }))->start();
```

Comment:
NetworkProcess::preconnectTo is doing things slightly differently, for instance:
- session->networkLoadScheduler().startedPreconnectForMainResource(url, userAgent);
- setTimeout(10_s);
- It does things in case ENABLE(ADDITIONAL_PRECONNECT_ON_HTTP_1X) is true
- It sets setIsAppInitiated
- It sets priority to very high (not sure this is what we want).

Reply:
Let me see what code we can share. The code I wrote is more similar to `SpeculativeLoadManager::preconnectForSubresource`.

Reply:
The code I wrote is more like `SpeculativeLoadManager::preconnectForSubresource`. That's a bit more like what I'm doing here, since `NetworkProcess::preconnectTo` is used for main resource preconnects. That's also why `NetworkProcess::preconnectTo` has some additional weirdness around ENABLE_ADDITIONAL_PRECONNECT_ON_HTTP_1X etc, notifying the resource scheduler that a main resource load is about to start, etc.

I'm going to defer trying to unify these preconnect implementations to https://bugs.webkit.org/show_bug.cgi?id=253176.